fix completion starting with paren + perf optimize

litoj opened the pull request:
In the current form, expressions starting with `(` are recognized as with unmatched parens, this can be fixed by simplifying the unmatched paren detector.

Also the pattern used to detect a keyword is static, so there is no reason to concatenate the string on every test. Neither do we need to escape the math functions - identifiers can't contains any regex-specific chars.

litoj committed:
fix completion starting with paren + perf optimize

litoj committed:
perf: simplify trigger chars

calc has no reason to trigger when writing the function name, only when the function has arguments and is closed can we use the content.

JosefLitos commented:
@hrsh7th could you please take a look and possibly point out some changes that should be made in order for this PR to be acceptable?

litoj committed:
revert removed space as trigger char

hrsh7th commented:
I think you implementation does not complete when the line is `(1 + 2|`

hrsh7th commented:
But your implementation is optimzied. Could you fix this? I hope to merge this :)

JosefLitos commented:
will take a look

litoj committed:
fix: always evaluate longest valid match

litoj committed:
perf: value after ',' isn't used; \s -> ' '

- removed `','` as a trigger char because the expression after
  it didn't get returned and the previous one stayed trailing
- keep trailing spaces - if there are any then the user
  probably wants to align something
- replace `\s` with `' '` - `'\t'` is for alignment ->
  user doesn't need evaluation, `'\n'` is never there

JosefLitos commented:
Could you please review the last update? I think it should be ready to merge.

hrsh7th commented:
Oh! sorry. I think it should be merged.

litoj commented:
No problem, btw I am currently working on another improvement to improve unmatched parenthesis handling (evaluating the largest valid range). So I may send another PR today.
